pkg/infra/game/stage/trade/trade.go

```diff
@@ -0,0 +1,241 @@
+		agentState.Shields = *commons.SliceToImmutableList(availableShields[agentID])
+		s.AgentState[agentID] = agentState
+	}
+}
+func HandleTradeMessage(agentID commons.ID, negotiation message.TradeMessage,
+	info *internal.Info,
+	agentState map[commons.ID]state.AgentState,
+) {
+	switch msg := negotiation.(type) {
+	case message.TradeAbstain:
+	case message.TradeResponse:
+		HandleTradeResponse(agentID, msg, info, agentState)
+	case message.TradeRequest:
+		HandleTradeRequest(agentID, msg, info)
+	}
+}
+
+func HandleTradeRequest(agentID commons.ID, msg message.TradeRequest,
+	info *internal.Info,
+) {
+	// add new negotiation to ongoing negotiations
+	negotiation := message.NewTradeNegotiation(agentID, msg.CounterPartyID, msg.Offer, msg.Demand)
+	info.Negotiations()[negotiation.Id] = negotiation
+	// remove offered item from available items
+	if msg.Offer.ItemType == commons.Weapon {
+		info.Weapons()[agentID] = RemoveItem(info.Weapons()[agentID], msg.Offer.Item)
+	} else {
+		info.Shields()[agentID] = RemoveItem(info.Shields()[agentID], msg.Offer.Item)
+	}
+}
+
+func HandleTradeResponse(agentID commons.ID, msg message.TradeResponse,
+	info *internal.Info,
+	agentState map[commons.ID]state.AgentState,
+) {
+	switch resp := msg.(type) {
+	case message.TradeAccept:
+		negotiation := info.Negotiations()[resp.TradeID]
+		if negotiation.Notarize(agentState) {
+			ExecuteTrade(&info.Inventory, negotiation)
+		}
+		RemoveFromNegotiation(resp.TradeID, agentID, info.Negotiations())
+	case message.TradeReject:
+		negotiation := info.Negotiations()[resp.TradeID]
+		RemoveFromNegotiation(resp.TradeID, agentID, info.Negotiations())
+		PutBackItems(&info.Inventory, negotiation)
+	case message.TradeBargain:
+		negotiation := info.Negotiations()[resp.TradeID]
+		if !negotiation.IsInvolved(agentID) {
+			break
+		}
+		// update ongoing negotiations
+		negotiation.UpdateDemand(agentID, resp.Demand)
+		oldOffer, replaceOffer := negotiation.UpdateOffer(agentID, resp.Offer)
+		info.Negotiations()[resp.TradeID] = negotiation
+		// update available items
+		if replaceOffer {
+			if oldOffer.ItemType == commons.Weapon {
+				info.Weapons()[agentID] = append(info.Weapons()[agentID], oldOffer.Item)
+				info.Weapons()[agentID] = RemoveItem(info.Weapons()[agentID], resp.Offer.Item)
+			} else {
+				info.Shields()[agentID] = append(info.Shields()[agentID], oldOffer.Item)
+				info.Shields()[agentID] = RemoveItem(info.Shields()[agentID], resp.Offer.Item)
+			}
+		}
+	}
+}
+
+func RemoveFromNegotiation(tradeID commons.TradeID, agentID commons.ID, negotiations map[commons.TradeID]message.TradeNegotiation) {
+	negotiation := negotiations[tradeID]
+	if negotiation.IsInvolved(agentID) {
+		delete(negotiations, tradeID)
+	}
+}
+
+func AddItem(available map[commons.ID][]state.Item, agentID commons.ID, item state.Item) {
+	availableList := available[agentID]
+	availableList = append(availableList, item)
```

Comment:
Given that previously we were continuously sorting the inventories such that agents knew what their strongest weapons were, should we consider using a `immutable.SortedMap[Item, struct{}]` in general for this? Maybe a future PR if we have time..

Reply:
Do you mean `immutable.SortedMap[commons.ItemID, struct{}]`?

Reply:
I was initially thinking that the other way is better since it’s a sorted set and we can sort by the actually value rather than the string, but maybe it’s not possible.

Reply:
yeah, it will be nice to sort items by values. I googled and also searched in immutable repo, didn't managed to find an easy solution to use struct as key.

Reply:
No worries. Maybe a future PR, not a big issue here.